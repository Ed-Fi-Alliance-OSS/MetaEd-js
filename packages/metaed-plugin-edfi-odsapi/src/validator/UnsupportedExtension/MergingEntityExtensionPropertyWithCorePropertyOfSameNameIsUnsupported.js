// @flow

// 2.x - METAED-711 - ODS-1732
import { asTopLevelEntity, getAllEntitiesOfType, versionSatisfies } from 'metaed-core';
import type {
  EntityProperty,
  MetaEdEnvironment,
  ModelBase,
  PluginEnvironment,
  SemVer,
  TopLevelEntity,
  ValidationFailure,
} from 'metaed-core';
import { collectSingleEntity, propertyCollector } from '../ValidatorShared/PropertyCollector';

const validatorName: string = 'MergingEntityExtensionPropertyWithCorePropertyOfSameNameIsUnsupported';
const targetTechnologyVersion: SemVer = '2.x';

function isTargetTechnologyVersion(metaEd: MetaEdEnvironment): boolean {
  return versionSatisfies(
    ((metaEd.plugin.get('edfiOdsApi'): any): PluginEnvironment).targetTechnologyVersion,
    targetTechnologyVersion,
  );
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  if (!isTargetTechnologyVersion(metaEd)) return failures;

  getAllEntitiesOfType(
    metaEd,
    'associationExtension',
    'associationSubclass',
    'domainEntityExtension',
    'domainEntitySubclass',
  )
    .map((entity: ModelBase) => asTopLevelEntity(entity))
    .forEach((entity: TopLevelEntity) => {
      if (!entity.namespace.isExtension || !entity.baseEntity) return;
      if (entity.baseEntity.namespace.isExtension) return;

      const baseEntityResult: {
        referencedEntities: Array<{
          withContext: string,
          entity: TopLevelEntity,
        }>,
        properties: Array<{
          withContext: string,
          property: EntityProperty,
        }>,
      } = collectSingleEntity(
        entity.baseEntity,
        true,
        (referencedEntity, property) => ({
          withContext: property.withContext,
          entity: asTopLevelEntity(referencedEntity),
        }),
        (referencedEntity, property) => ({
          withContext: property.withContext,
          property,
        }),
      );

      const extensionEntityResult: {
        referencedEntities: Array<{
          withContext: string,
          entity: TopLevelEntity,
        }>,
        properties: Array<{
          withContext: string,
          property: EntityProperty,
        }>,
      } = collectSingleEntity(
        entity,
        true,
        (referencedEntity, property) => ({
          withContext: property.withContext,
          entity: asTopLevelEntity(referencedEntity),
        }),
        (referencedEntity, property) => ({
          withContext: property.withContext,
          property,
        }),
      );

      const baseProperties: Array<{ withContext: string, property: EntityProperty }> = [];
      const extensionProperties: Array<{ withContext: string, property: EntityProperty }> = [];

      baseEntityResult.properties.forEach(result => {
        baseProperties.push({ withContext: result.withContext, property: result.property });
      });
      baseEntityResult.referencedEntities.forEach(result => {
        baseProperties.push(
          ...propertyCollector(result.entity).map(property => ({
            withContext: result.withContext,
            property,
          })),
        );
      });

      extensionEntityResult.properties.forEach(result => {
        baseProperties.push({ withContext: result.withContext, property: result.property });
      });
      extensionEntityResult.referencedEntities.forEach(referencedEntity => {
        extensionProperties.push(
          ...propertyCollector(referencedEntity.entity).map(property => ({
            withContext: referencedEntity.withContext,
            property,
          })),
        );
      });
      const extensionPropertyNames: Array<string> = extensionProperties.map(
        x => x.withContext + x.property.withContext + x.property.metaEdName,
      );

      const duplicateProperties: Array<{
        withContext: string,
        property: EntityProperty,
      }> = baseProperties.filter(x =>
        extensionPropertyNames.includes(x.withContext + x.property.withContext + x.property.metaEdName),
      );

      duplicateProperties.forEach((duplicate: { withContext: string, property: EntityProperty }) => {
        if (!entity.baseEntity) return;
        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-1732] ${entity.typeHumanizedName} ${
            entity.metaEdName
          } has a property path that conflicts with property ${duplicate.property.metaEdName} on core ${
            entity.baseEntity.typeHumanizedName
          } ${
            entity.baseEntity.metaEdName
          }. Merging properties of an extension entity with a core property of the same name is currently unsupported by the ODS/API.`,
          sourceMap: duplicate.property.sourceMap.type,
          fileMap: null,
        });
      });
    });
  return failures;
}
