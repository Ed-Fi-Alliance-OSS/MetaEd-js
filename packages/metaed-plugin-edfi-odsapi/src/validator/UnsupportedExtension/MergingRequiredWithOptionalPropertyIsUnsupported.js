// @flow

// 2.x - METAED-695 - ODS-1177
import { asTopLevelEntity, getAllTopLevelEntities } from 'metaed-core';
import type { EntityProperty, MetaEdEnvironment, ModelBase, TopLevelEntity, ValidationFailure } from 'metaed-core';
import { collectSingleEntity, propertyCollector } from '../ValidatorShared/PropertyCollector';

type PropertyCollectorArray = Array<{
  withContext: string,
  isOptional: boolean,
  property: EntityProperty,
}>;

const validatorName: string = 'MergingRequiredWithOptionalPropertyIsUnsupported';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  getAllTopLevelEntities(metaEd.entity)
    .map((entity: ModelBase) => asTopLevelEntity(entity))
    .forEach((entity: TopLevelEntity) => {
      const result: {
        referencedEntities: Array<{
          withContext: string,
          isOptional: boolean,
          entity: TopLevelEntity,
        }>,
        properties: PropertyCollectorArray,
      } = collectSingleEntity(
        entity,
        true,
        (referencedEntity, property) => ({
          withContext: property.withContext,
          isOptional: property.isOptional,
          entity: asTopLevelEntity(referencedEntity),
        }),
        (referencedEntity, property) => ({
          withContext: property.withContext,
          isOptional: property.isOptional,
          property,
        }),
      );

      const optionalProperties: PropertyCollectorArray = [];
      const requiredProperties: PropertyCollectorArray = [];

      requiredProperties.push(...result.properties.filter(property => !property.isOptional));
      optionalProperties.push(...result.properties.filter(property => property.isOptional));
      result.referencedEntities.forEach(referencedEntity => {
        if (referencedEntity.isOptional)
          optionalProperties.push(
            ...propertyCollector(referencedEntity.entity).map(property => ({
              withContext: referencedEntity.withContext,
              isOptional: referencedEntity.isOptional,
              property,
            })),
          );
        else {
          requiredProperties.push(
            ...propertyCollector(referencedEntity.entity).map(property => ({
              withContext: referencedEntity.withContext,
              isOptional: referencedEntity.isOptional,
              property,
            })),
          );
        }
      });

      const requiredPropertiesNames: Array<string> = requiredProperties.map(
        x => x.withContext + x.property.withContext + x.property.metaEdName,
      );
      const duplicateProperties: PropertyCollectorArray = optionalProperties.filter(x =>
        requiredPropertiesNames.includes(x.withContext + x.property.withContext + x.property.metaEdName),
      );

      duplicateProperties.forEach((duplicate: { withContext: string, isOptional: boolean, property: EntityProperty }) => {
        // skip failure message for now if this is a core only issue - in future maybe report if in a core-only mode (like IDE Alliance mode)
        if (!entity.namespaceInfo.isExtension) return;

        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-1177] ${entity.typeHumanizedName} ${
            entity.metaEdName
          } has both a required and optional property path to property ${duplicate.withContext +
            duplicate.property
              .metaEdName}. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API.`,
          sourceMap: duplicate.property.sourceMap.metaEdName,
          fileMap: null,
        });
        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-1177] ${entity.typeHumanizedName} ${
            entity.metaEdName
          } has both a required and optional property path to property ${duplicate.withContext +
            duplicate.property
              .metaEdName}. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API.`,
          sourceMap: entity.sourceMap.metaEdName,
          fileMap: null,
        });
      });
    });

  return failures;
}
