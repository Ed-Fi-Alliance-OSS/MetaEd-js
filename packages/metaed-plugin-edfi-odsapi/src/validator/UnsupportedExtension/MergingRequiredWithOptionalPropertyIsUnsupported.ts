// 2.x - METAED-695 - ODS-1177
import { asTopLevelEntity, getAllTopLevelEntitiesForNamespaces, versionSatisfies, V2Only } from '@edfi/metaed-core';
import {
  EntityProperty,
  MetaEdEnvironment,
  ModelBase,
  TopLevelEntity,
  ValidationFailure,
  SemVer,
  PluginEnvironment,
} from '@edfi/metaed-core';
import { collectSingleEntity, propertyCollector } from '../ValidatorShared/PropertyCollector';

type PropertyCollectorArray = {
  roleName: string;
  isOptional: boolean;
  property: EntityProperty;
}[];

const validatorName = 'MergingRequiredWithOptionalPropertyIsUnsupported';
const targetTechnologyVersion: SemVer = V2Only;

function isTargetTechnologyVersion(metaEd: MetaEdEnvironment): boolean {
  return versionSatisfies(
    (metaEd.plugin.get('edfiOdsApi') as PluginEnvironment).targetTechnologyVersion,
    targetTechnologyVersion,
  );
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  if (!isTargetTechnologyVersion(metaEd)) return failures;

  getAllTopLevelEntitiesForNamespaces([...metaEd.namespace.values()])
    .map((entity: ModelBase) => asTopLevelEntity(entity))
    .forEach((entity: TopLevelEntity) => {
      const result: {
        referencedEntities: {
          roleName: string;
          isOptional: boolean;
          entity: TopLevelEntity;
        }[];
        properties: PropertyCollectorArray;
      } = collectSingleEntity(
        entity,
        true,
        (referencedEntity, property) => ({
          roleName: property.roleName,
          isOptional: property.isOptional,
          entity: asTopLevelEntity(referencedEntity),
        }),
        (_referencedEntity, property) => ({
          roleName: property.roleName,
          isOptional: property.isOptional,
          property,
        }),
      );

      const optionalProperties: PropertyCollectorArray = [];
      const requiredProperties: PropertyCollectorArray = [];

      requiredProperties.push(...result.properties.filter((property) => !property.isOptional));
      optionalProperties.push(...result.properties.filter((property) => property.isOptional));
      result.referencedEntities.forEach((referencedEntity) => {
        if (referencedEntity.isOptional)
          optionalProperties.push(
            ...propertyCollector(referencedEntity.entity).map((property) => ({
              roleName: referencedEntity.roleName,
              isOptional: referencedEntity.isOptional,
              property,
            })),
          );
        else {
          requiredProperties.push(
            ...propertyCollector(referencedEntity.entity).map((property) => ({
              roleName: referencedEntity.roleName,
              isOptional: referencedEntity.isOptional,
              property,
            })),
          );
        }
      });

      const requiredPropertiesNames: string[] = requiredProperties.map(
        (x) => x.roleName + x.property.roleName + x.property.metaEdName,
      );
      const duplicateProperties: PropertyCollectorArray = optionalProperties.filter((x) =>
        requiredPropertiesNames.includes(x.roleName + x.property.roleName + x.property.metaEdName),
      );

      duplicateProperties.forEach((duplicate: { roleName: string; isOptional: boolean; property: EntityProperty }) => {
        // skip failure message for now if this is a core only issue - in future maybe report if in a core-only mode (like IDE Alliance mode)
        if (!entity.namespace.isExtension) return;

        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-1177] ${entity.typeHumanizedName} ${
            entity.metaEdName
          } has both a required and optional property path to property ${
            duplicate.roleName + duplicate.property.metaEdName
          }. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API.`,
          sourceMap: duplicate.property.sourceMap.metaEdName,
          fileMap: null,
        });
        failures.push({
          validatorName,
          category: 'warning',
          message: `[ODS-1177] ${entity.typeHumanizedName} ${
            entity.metaEdName
          } has both a required and optional property path to property ${
            duplicate.roleName + duplicate.property.metaEdName
          }. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API.`,
          sourceMap: entity.sourceMap.metaEdName,
          fileMap: null,
        });
      });
    });

  return failures;
}
