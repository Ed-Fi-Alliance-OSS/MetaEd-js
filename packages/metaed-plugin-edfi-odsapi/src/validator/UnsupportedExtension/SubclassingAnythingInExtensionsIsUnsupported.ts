// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

// METAED-1052
import { MetaEdEnvironment, ValidationFailure, TopLevelEntity } from '@edfi/metaed-core';
import { getAllEntitiesOfType } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  (getAllEntitiesOfType(metaEd, 'domainEntitySubclass', 'associationSubclass') as TopLevelEntity[]).forEach(
    (entitySubclass: TopLevelEntity) => {
      if (!entitySubclass.baseEntity) return;
      if (!entitySubclass.namespace.isExtension) return;
      failures.push({
        validatorName: 'SubclassingAnythingInExtensionsIsUnsupported',
        category: 'error',
        message: `Subclasses are currently unsupported by the ODS/API in extension projects.`,
        sourceMap: entitySubclass.sourceMap.metaEdName,
        fileMap: null,
      });
    },
  );

  return failures;
}
