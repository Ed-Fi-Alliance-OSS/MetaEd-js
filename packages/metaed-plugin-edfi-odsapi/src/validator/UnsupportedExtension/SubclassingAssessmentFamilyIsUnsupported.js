// @flow

// 3.1.X.7 - METAED-701
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.domainEntitySubclass.forEach(domainEntitySubclass => {
    if (!domainEntitySubclass.baseEntity) return;
    if (!domainEntitySubclass.namespaceInfo.isExtension) return;
    if (domainEntitySubclass.baseEntityName === 'AssessmentFamily') {
      failures.push({
        validatorName: 'SubclassingAssessmentFamilyIsUnsupported',
        category: 'warning',
        message: `${domainEntitySubclass.typeHumanizedName} ${
          domainEntitySubclass.metaEdName
        } is an AssessmentFamily subclass.  AssessmentFamily subclasses are currently unsupported by the ODS/API.`,
        sourceMap: domainEntitySubclass.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}
