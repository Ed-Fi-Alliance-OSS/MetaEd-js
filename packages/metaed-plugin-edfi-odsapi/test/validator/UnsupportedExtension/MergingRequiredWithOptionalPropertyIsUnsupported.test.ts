import {
  newMetaEdEnvironment,
  newPluginEnvironment,
  MetaEdTextBuilder,
  DomainEntityBuilder,
  NamespaceBuilder,
} from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { initialize as initializeUnifiedPlugin } from '@edfi/metaed-plugin-edfi-unified';
import { initializeNamespaceDependencies } from '../ValidationTestHelper';
import { validate } from '../../../src/validator/UnsupportedExtension/MergingRequiredWithOptionalPropertyIsUnsupported';

describe('when a domain entity has both a required and optional reference to a property of the same name but is in core', (): void => {
  let failures: ValidationFailure[];

  beforeAll(() => {
    const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
    const domainEntityName1 = 'DomainEntityName1';
    const domainEntityName2 = 'DomainEntityName2';
    const domainEntityName3 = 'DomainEntityName3';

    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IntegerPropertyName1', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName1, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName3)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName2, 'Documentation')
      .withDomainEntityProperty(domainEntityName1, 'Documentation', false, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    initializeUnifiedPlugin().enhancer.forEach((enhance) => enhance(metaEd));
    metaEd.plugin.set(
      'edfiOdsApi',
      Object.assign(newPluginEnvironment(), {
        targetTechnologyVersion: '2.0.0',
      }),
    );
    failures = validate(metaEd);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when a domain entity has both a required and optional reference to a property of the same name in an extension', (): void => {
  let failures: ValidationFailure[];

  beforeAll(() => {
    const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
    const domainEntityName1 = 'DomainEntityName1';
    const domainEntityName2 = 'DomainEntityName2';
    const domainEntityName3 = 'DomainEntityName3';

    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IntegerPropertyName1', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName1, 'Documentation')
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartDomainEntity(domainEntityName3)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(`EdFi.${domainEntityName2}`, 'Documentation')
      .withDomainEntityProperty(`EdFi.${domainEntityName1}`, 'Documentation', false, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    initializeNamespaceDependencies(metaEd, 'EdFi', 'Extension');
    initializeUnifiedPlugin().enhancer.forEach((enhance) => enhance(metaEd));
    metaEd.plugin.set(
      'edfiOdsApi',
      Object.assign(newPluginEnvironment(), {
        targetTechnologyVersion: '2.0.0',
      }),
    );
    failures = validate(metaEd);
  });

  it('should have validation failure', (): void => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('MergingRequiredWithOptionalPropertyIsUnsupported');
    expect(failures[0].category).toBe('warning');
    expect(failures[0].message).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> message',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> sourceMap',
    );
    expect(failures[1].validatorName).toBe('MergingRequiredWithOptionalPropertyIsUnsupported');
    expect(failures[1].category).toBe('warning');
    expect(failures[1].message).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> message',
    );
    expect(failures[1].sourceMap).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> sourceMap',
    );
  });
});

describe('when a domain entity has both a required and optional reference to domain entity with a property of the same name', (): void => {
  let failures: ValidationFailure[];

  beforeAll(() => {
    const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
    const domainEntityName1 = 'DomainEntityName1';
    const domainEntityName2 = 'DomainEntityName2';
    const domainEntityName3 = 'DomainEntityName3';
    const domainEntityName4 = 'DomainEntityName4';
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IntegerPropertyName1', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName1, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName3)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName1, 'Documentation')
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartDomainEntity(domainEntityName4)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(`EdFi.${domainEntityName2}`, 'Documentation')
      .withDomainEntityProperty(`EdFi.${domainEntityName3}`, 'Documentation', false, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    initializeNamespaceDependencies(metaEd, 'EdFi', 'Extension');
    initializeUnifiedPlugin().enhancer.forEach((enhance) => enhance(metaEd));
    metaEd.plugin.set(
      'edfiOdsApi',
      Object.assign(newPluginEnvironment(), {
        targetTechnologyVersion: '2.0.0',
      }),
    );
    failures = validate(metaEd);
  });

  it('should have validation failure', (): void => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('MergingRequiredWithOptionalPropertyIsUnsupported');
    expect(failures[0].category).toBe('warning');
    expect(failures[0].message).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> message',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> sourceMap',
    );
    expect(failures[1].validatorName).toBe('MergingRequiredWithOptionalPropertyIsUnsupported');
    expect(failures[1].category).toBe('warning');
    expect(failures[1].message).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> message',
    );
    expect(failures[1].sourceMap).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> sourceMap',
    );
  });
});

describe('when a domain entity has both a required and optional reference to a property of the same name with same context name', (): void => {
  let failures: ValidationFailure[];

  beforeAll(() => {
    const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
    const contextName = 'ContextName';
    const domainEntityName1 = 'DomainEntityName1';
    const domainEntityName2 = 'DomainEntityName2';
    const domainEntityName3 = 'DomainEntityName3';
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IntegerPropertyName1', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName1, 'Documentation')
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartDomainEntity(domainEntityName3)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(`EdFi.${domainEntityName2}`, 'Documentation', contextName)
      .withDomainEntityProperty(`EdFi.${domainEntityName1}`, 'Documentation', false, false, false, contextName)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    initializeNamespaceDependencies(metaEd, 'EdFi', 'Extension');
    initializeUnifiedPlugin().enhancer.forEach((enhance) => enhance(metaEd));
    metaEd.plugin.set(
      'edfiOdsApi',
      Object.assign(newPluginEnvironment(), {
        targetTechnologyVersion: '2.0.0',
      }),
    );
    failures = validate(metaEd);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('MergingRequiredWithOptionalPropertyIsUnsupported');
    expect(failures[0].category).toBe('warning');
    expect(failures[0].message).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> message',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> sourceMap',
    );
    expect(failures[1].validatorName).toBe('MergingRequiredWithOptionalPropertyIsUnsupported');
    expect(failures[1].category).toBe('warning');
    expect(failures[1].message).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> message',
    );
    expect(failures[1].sourceMap).toMatchSnapshot(
      'when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> sourceMap',
    );
  });
});

describe('when a domain entity has both a required and optional reference to a property of the same name with different context name', (): void => {
  let failures: ValidationFailure[];

  beforeAll(() => {
    const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
    const domainEntityName1 = 'DomainEntityName1';
    const domainEntityName2 = 'DomainEntityName2';
    const domainEntityName3 = 'DomainEntityName3';
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IntegerPropertyName1', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName1, 'Documentation')
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartDomainEntity(domainEntityName3)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(`EdFi.${domainEntityName2}`, 'Documentation')
      .withDomainEntityProperty(`EdFi.${domainEntityName1}`, 'Documentation', false, false, false, 'ContextName')
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    initializeNamespaceDependencies(metaEd, 'EdFi', 'Extension');
    initializeUnifiedPlugin().enhancer.forEach((enhance) => enhance(metaEd));
    metaEd.plugin.set(
      'edfiOdsApi',
      Object.assign(newPluginEnvironment(), {
        targetTechnologyVersion: '2.0.0',
      }),
    );
    failures = validate(metaEd);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when a domain entity has both a required and optional reference to a property of the same name with different context name on property', (): void => {
  let failures: ValidationFailure[];

  beforeAll(() => {
    const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
    const domainEntityName1 = 'DomainEntityName1';
    const domainEntityName2 = 'DomainEntityName2';
    const IntegerPropertyName1 = 'IntegerPropertyName1';

    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity(IntegerPropertyName1, 'Documentation', '100', '0', 'ContextName')
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(`EdFi.${domainEntityName1}`, 'Documentation')
      .withIntegerProperty(IntegerPropertyName1, 'Documentation', false, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    initializeNamespaceDependencies(metaEd, 'EdFi', 'Extension');
    initializeUnifiedPlugin().enhancer.forEach((enhance) => enhance(metaEd));
    metaEd.plugin.set(
      'edfiOdsApi',
      Object.assign(newPluginEnvironment(), {
        targetTechnologyVersion: '2.0.0',
      }),
    );
    failures = validate(metaEd);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});
