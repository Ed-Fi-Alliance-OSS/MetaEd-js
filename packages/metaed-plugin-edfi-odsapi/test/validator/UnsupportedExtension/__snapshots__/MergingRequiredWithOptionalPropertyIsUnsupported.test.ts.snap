// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when a domain entity has both a required and optional reference to a property of the same name in an extension should have validation failure: when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> message 1`] = `"[ODS-1177] Domain Entity DomainEntityName3 has both a required and optional property path to property IntegerPropertyName1. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API."`;

exports[`when a domain entity has both a required and optional reference to a property of the same name in an extension should have validation failure: when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> message 2`] = `"[ODS-1177] Domain Entity DomainEntityName3 has both a required and optional property path to property IntegerPropertyName1. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API."`;

exports[`when a domain entity has both a required and optional reference to a property of the same name in an extension should have validation failure: when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> sourceMap 1`] = `
Object {
  "column": 12,
  "line": 5,
  "tokenText": "IntegerPropertyName1",
}
`;

exports[`when a domain entity has both a required and optional reference to a property of the same name in an extension should have validation failure: when a domain entity has both a required and optional reference to a property of the same name should have validation failure -> sourceMap 2`] = `
Object {
  "column": 16,
  "line": 18,
  "tokenText": "DomainEntityName3",
}
`;

exports[`when a domain entity has both a required and optional reference to a property of the same name with same context name should have no validation failures: when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> message 1`] = `"[ODS-1177] Domain Entity DomainEntityName3 has both a required and optional property path to property ContextNameIntegerPropertyName1. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API."`;

exports[`when a domain entity has both a required and optional reference to a property of the same name with same context name should have no validation failures: when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> message 2`] = `"[ODS-1177] Domain Entity DomainEntityName3 has both a required and optional property path to property ContextNameIntegerPropertyName1. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API."`;

exports[`when a domain entity has both a required and optional reference to a property of the same name with same context name should have no validation failures: when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> sourceMap 1`] = `
Object {
  "column": 12,
  "line": 5,
  "tokenText": "IntegerPropertyName1",
}
`;

exports[`when a domain entity has both a required and optional reference to a property of the same name with same context name should have no validation failures: when a domain entity has both a required and optional reference to a property of the same name with same context name should have validation failure -> sourceMap 2`] = `
Object {
  "column": 16,
  "line": 18,
  "tokenText": "DomainEntityName3",
}
`;

exports[`when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure: when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> message 1`] = `"[ODS-1177] Domain Entity DomainEntityName4 has both a required and optional property path to property IntegerPropertyName1. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API."`;

exports[`when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure: when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> message 2`] = `"[ODS-1177] Domain Entity DomainEntityName4 has both a required and optional property path to property IntegerPropertyName1. Merging required properties with optional properties of the same name is currently unsupported by the ODS/API."`;

exports[`when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure: when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> sourceMap 1`] = `
Object {
  "column": 12,
  "line": 5,
  "tokenText": "IntegerPropertyName1",
}
`;

exports[`when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure: when a domain entity has both a required and optional reference to domain entity with a property of the same name should have validation failure -> sourceMap 2`] = `
Object {
  "column": 16,
  "line": 25,
  "tokenText": "DomainEntityName4",
}
`;
